### ISSUE TITLE
Cant Interact with fire near fences

### ISSUE THREAD

**Issue by FabioZumbi12** [PaperMC/Paper#1915]
### What behaviour is expected:
Deny interact with fire using the event `PlayerInteractEvent` and `e.getPlayer().getTargetBlock(null, 5);` if the fire is near or behind a fence. Seems the hitbox of fence is bypassing the hit, the fire still getting the hit too and extinguishing him.

### What behaviour is observed:
Using the event `PlayerInteractEvent`, when interact with fire behind fences using the following code:
```Java
    @EventHandler
    public void onInteractFire(PlayerInteractEvent e) {
        Block b = e.getPlayer().getTargetBlock(null, 5);
        RedProtect.get().logger.severe("Block Type: " + b.getType().name());
    }
```
The output:
![image](https://user-images.githubusercontent.com/5833818/54891204-5f6e5e00-4e8b-11e9-977a-0dfe58c16406.png)

The material returned is the fence, and the fire is extinguished.
![image](https://user-images.githubusercontent.com/5833818/54891166-3d74db80-4e8b-11e9-9181-fd0e8cebf3d7.png)

### Steps/models to reproduce:
Just put fire, and in the front of fire put any type of fence.

### Plugin list:
CratesPlus, Essentials*, EssentialsGeoIP*, HolographicDisplays, LuckPerms, Multiverse-Core*, Multiverse-EssentialsSupport*, PlaceholderAPI*, ProtocolLib*, RankUpper, RedProtect, UltimateChat, Vault*, Votifier*, WorldBorder*, WorldEdit

### Paper build number:
This server is running Paper version git-Paper-582 (MC: 1.13.2) (Implementing API version 1.13.2-R0.1-SNAPSHOT)

**Comment by aikar** [comment]
This sounds like expected behavior. You need to use the getTargetBlockExact method to get the behavior you desire which is for this exact purpose.

**Comment by FabioZumbi12** [comment]
Tryed this way and return only NetherRack and sand, not the fire:

```Java
    @EventHandler
    public void onInteractFire(PlayerInteractEvent e) {
        Block b = e.getPlayer().getTargetBlockExact(5);
        RedProtect.get().logger.severe("Block: " + b.getType().name());
    }
```

![image](https://user-images.githubusercontent.com/5833818/54892503-8ed39980-4e90-11e9-9ba2-b9445fba788d.png)

**Comment by aikar** [comment]
youll want to do something with ray tracing, find target face interacted, find opposite block, if fire, cancel.

come on Discord and ask in #paper-dev and Billy may be able to help you, he did all the ray tracing API's (but that api is also in Bukkit directly now)

**Comment by FabioZumbi12** [comment]
OK, i've joined discord now!

**Comment by BillyGalbreath** [comment]
Press F3 in your client and tell me what block shows on the right when looking at the fire? The raytracing stuff doesn't see blocks it sees collision boxes, which fire has none of.

**Comment by FabioZumbi12** [comment]
Hi Billy, here is what i got:

Through fence hitbox:
![image](https://user-images.githubusercontent.com/5833818/54971749-b432d700-4f66-11e9-8ddb-31c0baa33bd2.png)

The same aiming directly on fire:
![image](https://user-images.githubusercontent.com/5833818/54971767-d3316900-4f66-11e9-85df-8f0584a5d5f5.png)

The issue seems to be only with `e.getPlayer().getTargetBlock(null, 5);` and only this method get the fire block and only aiming directly, not through fence hitbox.

**Comment by aikar** [comment]
Since this is ultimately plugin development help, please ask on Discord.

ultimately I think you are looking for this:
```java
TargetBlockInfo targetBlockInfo = player.getTargetBlockInfo(5);
Block block = targetBlockInfo != null ? targetBlockInfo.getRelativeBlock() : null;
```

The API is working as intended with everything you've said so far, so please ask on Discord for any further help. but you'll want TargetBlockInfo in SOME fashion.

**Comment by FabioZumbi12** [comment]
I just want to report about `Player.getTargetBlock(null, 5);` exploit and theres no other listener to listen for interactions with fire (as i know and with what i tryed).

I didnt found a way to deny players to interact with FIRE than this way, the unique problem is when the player hit the fire near other "not full block" like fence os slab, the block returned by `Player.getTargetBlock(null, 5);` is the front block, not the fire, but fire is extinguished.

If i cant detect fire with interact listener and theres no way to fix getTargetBlock, here is a suggestion then.

Edit:
```Java
    @EventHandler
    public void onInteractFire(PlayerInteractEvent e) {
        if (e.getClickedBlock() != null){
            Block b = e.getClickedBlock().getRelative(e.getBlockFace());
            if (b.getType().equals(Material.FIRE)) {
                ...
            }
        }
    }
```
This supply my needs!

**Comment by BillyGalbreath** [comment]
@FabioZumbi12 you misunderstand.

`getTargetBlock` uses a BlockIterator internally, which is only as precise as 1x1x1 block. It does not take into consideration the collision boxes of the blocks.

This was a problem for me dating as far back as I can remember. I found some workaround solutions like [this one](https://www.spigotmc.org/threads/get-the-correct-sign-player-is-looking-at.179373/), but they never fully satisfied me.

Fast forward to late last year, I found a proper solution by exposing Mojang's raytrace stuff (as seen in F3). This is [PR #1410](https://github.com/PaperMC/Paper/pull/1410) where I added new getTargetBlock methods which do take into consideration collision boxes.

Best part yet, it actually can give you [the blockface being looked at](https://papermc.io/javadocs/1.13/com/destroystokyo/paper/block/TargetBlockInfo.html#getBlockFace--) so you can get the neighboring block. This is what you're after if you want to cancel clicking out fire.

Put simply, when you click fire Mojangs code detects the block you're looking at through the fire. It then checks the block next to the face of the block you clicked on the side you're looking at. If it's fire, then it puts it out. You're going to have to do that same check to see if it's fire.
